c2h/generators_vector.cu

```diff
@@ -0,0 +1,195 @@
+    random_to_vec_item_t(T min, T max, float* in, T* out)                          \
+        : m_min(min)                                                               \
+        , m_max(max)                                                               \
+        , m_in(in)                                                                 \
+        , m_out(out)                                                               \
+    {}                                                                             \
+    T m_min;                                                                       \
+    T m_max;                                                                       \
+    float* m_in{};                                                                 \
+    T* m_out{};                                                                    \
+  }
+
+RANDOM_TO_VEC_ITEM_SPEC(0, x);
+RANDOM_TO_VEC_ITEM_SPEC(1, y);
+RANDOM_TO_VEC_ITEM_SPEC(2, z);
+RANDOM_TO_VEC_ITEM_SPEC(3, w);
+#undef RANDOM_TO_VEC_ITEM_SPEC
+
+#if THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
+template <typename T, int VecItem>
+struct vec_gen_helper_t;
+
+template <typename T>
+struct vec_gen_helper_t<T, -1>
+{
+  static void gen(c2h::device_vector<T>&, T, T) {}
+};
+
+template <typename T, int VecItem>
+struct vec_gen_helper_t
+{
+  static void gen(c2h::device_vector<T>& data, T min, T max)
+  {
+    thrust::counting_iterator<size_t> cnt_begin(0);
+    thrust::counting_iterator<size_t> cnt_end(data.size());
+
+    generator_t& generator = generator_t::instance();
+    float* d_in            = generator.distribution();
+    T* d_out               = thrust::raw_pointer_cast(data.data());
+
+    generator.generate();
+
+    thrust::for_each(c2h::device_policy, cnt_begin, cnt_end, random_to_vec_item_t<T, VecItem>{min, max, d_in, d_out});
+
+    vec_gen_helper_t<T, VecItem - 1>::gen(data, min, max);
+  }
+};
+
+#  define VEC_SPECIALIZATION(TYPE, SIZE)                                                                     \
+    template <>                                                                                              \
+    void gen<TYPE##SIZE>(seed_t seed, c2h::device_vector<TYPE##SIZE> & data, TYPE##SIZE min, TYPE##SIZE max) \
+    {                                                                                                        \
+      generator_t& generator = generator_t::instance();                                                      \
+      generator.prepare_random_generator(seed, data.size());                                                 \
+      vec_gen_helper_t<TYPE##SIZE, SIZE - 1>::gen(data, min, max);                                           \
+    }
+
+VEC_SPECIALIZATION(char, 2);
+VEC_SPECIALIZATION(char, 3);
+VEC_SPECIALIZATION(char, 4);
+
+// VEC_SPECIALIZATION(uchar, 2);
+VEC_SPECIALIZATION(uchar, 3);
+// VEC_SPECIALIZATION(uchar, 4);
+
+VEC_SPECIALIZATION(short, 2);
+VEC_SPECIALIZATION(short, 3);
+VEC_SPECIALIZATION(short, 4);
```

Comment:
Remark: we may want to split this file a few times more to accelerate compilation, but this can be done in a separate PR later.